### PR TITLE
Patient caution enhancement and fixation

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR : 

- Fix Caution receipt 
- Fix Financial status of a patient 
- Add caution message in the cash and patient invoice modules

### CODE REVIEW

File: client/src/js/services/PatientService.js
Comment (mbayopanda) [inline client/src/js/services/PatientService.js:69]: The uuid is not patient's uuid but is the debtor_uuid of the patient

File: client/src/js/components/bhLatestInvoice.js
Comment (jniles) [inline client/src/js/components/bhLatestInvoice.js]: Should this have an error handler like `Notify`?

File: client/src/partials/cash/cash.js
Comment (jniles) [inline client/src/partials/cash/cash.js]: This is a good idea, but I think this might need to be warning instead of an error.  The patient can have a positive balance with the hospital, but it should be up to the patient if they want to use their caution to pay a particular invoice.

What do you think?

File: client/src/partials/patient_invoice/patientInvoice.html
Comment (jniles) [inline client/src/partials/patient_invoice/patientInvoice.html]: I think this `currency` filter should take the enterprise currency ID as a parameter.

File: client/src/partials/templates/bhLatestInvoice.tmpl.html
Comment (jniles) [inline client/src/partials/templates/bhLatestInvoice.tmpl.html]: All these currency filters should use the enterprise currency ID.

File: server/controllers/finance/debtors/index.js
Comment (jniles) [inline server/controllers/finance/debtors/index.js]: :+1:

File: server/controllers/finance/patient.js
Comment (jniles) [inline server/controllers/finance/patient.js:22]: Nice reuse of code!

File: server/controllers/finance/patient.js
Comment (jniles) [inline server/controllers/finance/patient.js]: You can remove these extra imports since they are never used.

File: server/controllers/finance/reports/financial.patient.js
Comment (jniles) [inline server/controllers/finance/reports/financial.patient.js]: Technically, the jsdoc `@overview` tag is for the file name.  This is fine, but might be confusing to someone who works with JSDoc a lot.

[Source](http://usejsdoc.org/tags-file.html).

File: server/controllers/finance/reports/index.js
Comment (jniles) [inline server/controllers/finance/reports/index.js]: I know you didn't write this, but `financialPatient` is a bit of a confusing name.  Maybe `patientFinancialStatus`?  Does it give the account number of the patient, or just return the balance, debit, credits to the patient?

**Review by jniles — CHANGES_REQUESTED** [review]
@mbayopanda, the code looks good.

Remember that anytime you use the `currency` filter, you should pass it an ID.  It defaults to USD, which will break if an enterprise decides to use anything other than USD as their enterprise currency.

I'm a bit concerned about making a positive balance block cash payments ... it seems to me the application should _suggest_ that an accountant use the positive debtor balance to pay the invoices, but that is really the debtor's decision and should be done with a journal voucher somewhere else.  What do you think?

File: client/src/partials/patient_invoice/patientInvoice.html
Comment (mbayopanda) [reply client/src/partials/patient_invoice/patientInvoice.html]: Invoices are always paid in $, but you're right because the enterprise currency is not necessary in $.

File: server/controllers/finance/reports/index.js
Comment (mbayopanda) [reply server/controllers/finance/reports/index.js]: 👍

File: client/src/partials/cash/cash.js
Comment (mbayopanda) [reply client/src/partials/cash/cash.js]: We can add a confirm modal, if the client want to pay for an invoice he must accept in the confirm modal box which  display a clear message about the caution and warning.

File: client/src/partials/templates/bhLatestInvoice.tmpl.html
Comment (mbayopanda) [reply client/src/partials/templates/bhLatestInvoice.tmpl.html]: 👍

**Comment by jniles** [comment]
@mbayopanda, when you are ready for a second review, feel free to ping me.

**Comment by mbayopanda** [comment]
@jniles i'm ready for a second review

**Comment by jniles** [comment]
@sfount, can you do this review to balance things out?

File: client/src/partials/cash/cash.html
Comment (sfount) [inline client/src/partials/cash/cash.html]: This is complicated logic for a template to be processing. 

Is it possible for this logic to be performed in the controller and then referenced by the template. It could then only be updated when the patient balance or cash slip is changed. 
The only real problem here is that it is difficult to reason about what the template is actually displaying here.

File: client/src/partials/cash/cash.js
Comment (sfount) [inline client/src/partials/cash/cash.js]: I think this is equivalent to

``` js
var hasCaution = vm.slip && vm.patientBalance < 0;
```

**Review by sfount — COMMENTED** [review]
This code LGTM. I have a few comments on template logic vs. controller however I don't think any of this needs to be blocking.

File: server/controllers/finance/reports/financial.patient.js
Comment (jniles) [inline server/controllers/finance/reports/financial.patient.js:30]: The actual route is called `financialPatient`.  Make sure your comments are up to date or they will confuse developers!

**Review by jniles — APPROVED** [review]

**Comment by jniles** [comment]
@mbayopanda, these changes LGTM. :+1:  As soon as your fix the merge conflicts, this will be merged.

**Pull request merged**
